packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,402 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:go_router/go_router.dart';
+
+final GlobalKey<NavigatorState> _sectionANavigatorKey =
+    GlobalKey<NavigatorState>(debugLabel: 'sectionANav');
+final GlobalKey<NavigatorState> _sectionBNavigatorKey =
+    GlobalKey<NavigatorState>(debugLabel: 'sectionBNav');
+
+// This example demonstrates how to setup nested navigation using a
+// BottomNavigationBar, where each tab uses its own persistent navigator, i.e.
+// navigation state is maintained separately for each tab. This setup also
+// enables deep linking into nested pages.
+//
+// This example demonstrates how to display routes within a ShellRoute using a
+// `nestedNavigationBuilder`. Navigators for the tabs ('Section A' and
+// 'Section B') are created via nested ShellRoutes. Note that no navigator will
+// be created by the "top" ShellRoute. This example is similar to the ShellRoute
+// example, but differs in that it is able to maintain the navigation state of
+// each tab.
+
+void main() {
+  runApp(NestedTabNavigationExampleApp());
+}
+
+/// Extension to get the child widget of the Page. Wish there were a better way...
+extension PageWithChild on Page<dynamic> {
+  /// Gets the child of the page
+  Widget? findTheChild() {
+    final Page<dynamic> widget = this;
+    if (widget is MaterialPage) {
+      return widget.child;
+    } else if (widget is CupertinoPage) {
+      return widget.child;
+    } else if (widget is CustomTransitionPage) {
+      return widget.child;
+    }
+    return null;
+    // An unsafer option... :
+    //return (this as dynamic).child as Widget;
+  }
+}
+
+/// ShellRoute that uses a bottom tab navigation (ScaffoldWithNavBar) with
+/// separate navigators for each tab.
+class BottomTabBarShellRoute extends ShellRoute {
```

Comment:
It will be good if we can provide something like this as a public API in go_router so that the developer don't have to manage the index stack and navigator by themselves. That's also part of the requirements of https://github.com/flutter/flutter/issues/99124

Reply:
If we do provide this, we need to think of a way to decouple BottomTabBar and scaffold from the IndexStack. The API itself should only focus on using index stack to preserve state and allow developer to build things around them.

Reply:
@chunhtai that would indeed be beneficial as currently there is a lot of code in the example that would have to be copy-pasted across projects.

Also, `BottomTabBarShellRoute` assumes that users would use this with bottom navigation only. But really, this is some kind of `MultiTabsShellRoute` that could work for bottom navigation, drawer navigation, navigation rail etc.

Reply:
Yes, I agree. I initially felt I didn't want to "pollute" go_router with too much special case implementation details, however I agree that at least a route providing the stacking behaviour could be beneficial. Looking at that right now.

Reply:
Solved with new widget class StackedNavigationScaffold (will possibly rename it to StackedNavigationShell).